File: animated-gif-lite/src/main/java/com/facebook/animated/giflite/draw/MovieScaleHolder.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.animated.giflite.draw;

class MovieScaleHolder {

  private final int mMovieWidth;
  private final int mMovieHeight;
  private int mViewPortWidth;
  private int mViewPortHeight;
  private float mScale = 1f;
  private float mLeft = 0;
  private float mTop = 0;

  public MovieScaleHolder(int movieWidth, int movieHeight) {
    mMovieWidth = movieWidth;
    mMovieHeight = movieHeight;
  }

  public synchronized float getScale() {
    return mScale;
  }

  public synchronized float getLeft() {
    return mLeft;
  }

  public synchronized float getTop() {
    return mTop;
  }

  public synchronized void updateViewPort(int viewPortWidth, int viewPortHeight) {
    if (mViewPortWidth == viewPortWidth && mViewPortHeight == viewPortHeight) {
      return;
    }
    mViewPortWidth = viewPortWidth;
    mViewPortHeight = viewPortHeight;
    determineScaleAndPosition();
  }

  private synchronized void determineScaleAndPosition() {

    float inputRatio = mMovieWidth / mMovieHeight;
    float outputRatio = mViewPortWidth / mViewPortHeight;

    int width = mViewPortWidth;
    int height = mViewPortHeight;
    if (outputRatio > inputRatio) {
      // Not enough width to fill the output. (Black bars on left and right.)
      width = (int) (mViewPortHeight * inputRatio);
    } else if (outputRatio < inputRatio) {
      // Not enough height to fill the output. (Black bars on top and bottom.)
      height = (int) (mViewPortWidth / inputRatio);
    }

    if (mViewPortWidth > mMovieWidth) {
      mScale = mMovieWidth / (float) mViewPortWidth;
    } else if (mMovieWidth > mViewPortWidth) {
      mScale = mViewPortWidth / (float) mMovieWidth;
    } else {
      mScale = 1f;
    }

    mLeft = ((mViewPortWidth - width) / 2f) / mScale;
    mTop = ((mViewPortHeight - height) / 2f) / mScale;
  }
}

File: animated-gif-lite/src/main/java/com/facebook/animated/giflite/draw/MovieScaleHolder.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.animated.giflite.draw

internal class MovieScaleHolder(private val movieWidth: Int, private val movieHeight: Int) {

  private var viewPortWidth = 0
  private var viewPortHeight = 0

  @get:Synchronized
  var scale = 1f
    private set

  @get:Synchronized
  var left = 0f
    private set

  @get:Synchronized
  var top = 0f
    private set

  @Synchronized
  fun updateViewPort(viewPortWidth: Int, viewPortHeight: Int) {
    if (this.viewPortWidth == viewPortWidth && this.viewPortHeight == viewPortHeight) {
      return
    }
    this.viewPortWidth = viewPortWidth
    this.viewPortHeight = viewPortHeight
    determineScaleAndPosition()
  }

  @Synchronized
  private fun determineScaleAndPosition() {
    val inputRatio = (movieWidth / movieHeight).toFloat()
    val outputRatio = (viewPortWidth / viewPortHeight).toFloat()
    var width = viewPortWidth
    var height = viewPortHeight
    if (outputRatio > inputRatio) {
      // Not enough width to fill the output. (Black bars on left and right.)
      width = (viewPortHeight * inputRatio).toInt()
    } else if (outputRatio < inputRatio) {
      // Not enough height to fill the output. (Black bars on top and bottom.)
      height = (viewPortWidth / inputRatio).toInt()
    }
    if (viewPortWidth > movieWidth) {
      scale = movieWidth / viewPortWidth.toFloat()
    } else if (movieWidth > viewPortWidth) {
      scale = viewPortWidth / movieWidth.toFloat()
    } else {
      scale = 1f
    }
    left = (viewPortWidth - width) / 2f / scale
    top = (viewPortHeight - height) / 2f / scale
  }
}
